phoenix-core/src/main/java/org/apache/phoenix/expression/LiteralExpression.java

```diff
@@ -57,181 +56,236 @@
-        else if (value instanceof Boolean) {
-            return getBooleanLiteralExpression((Boolean)value, determinism);
+
+        public Builder setSortOrder(SortOrder sortOrder) {
+            this.sortOrder=sortOrder;
+            return this;
         }
-        PDataType actualType = PDataType.fromLiteral(value);
-        type = type == null ? actualType : type;
-        try {
-            value = type.toObject(value, actualType);
-        } catch (IllegalDataException e) {
-            throw TypeMismatchException.newException(type, actualType, value.toString());
+
+        public Builder setByteValue(byte[] byteValue) {
+            this.byteValue=byteValue;
+            return this;
         }
-        byte[] b = type.isArrayType() ? ((PArrayDataType)type).toBytes(value, PArrayDataType.arrayBaseType(type), sortOrder, rowKeyOrderOptimizable) :
-                type.toBytes(value, sortOrder);
-        if (type == PVarchar.INSTANCE || type == PChar.INSTANCE) {
-            if (type == PChar.INSTANCE && maxLength != null  && b.length < maxLength) {
-                if (rowKeyOrderOptimizable) {
-                    b = type.pad(b, maxLength, sortOrder);
-                } else {
-                    b = StringUtil.padChar(b, maxLength);
-                }
-            } else if (value != null) {
-                maxLength = ((String)value).length();
+
+        private void setDefaults() {
+            if (this.determinism == null) {
+                this.determinism = Determinism.ALWAYS;
+            }
+            if (this.sortOrder == null) {
+                this.sortOrder = SortOrder.getDefault();
+            }
+            if (this.rowKeyOrderOptimizable == null) {
+                this.rowKeyOrderOptimizable = true;
             }
-        } else if (type.isArrayType()) {
-            maxLength = ((PhoenixArray)value).getMaxLength();
-        }
-        if (b.length == 0) {
-            return getTypedNullLiteralExpression(type, determinism);
-        }
-        if (maxLength == null) {
-            maxLength = type.isFixedWidth() ? type.getMaxLength(value) : null;
         }
-        return new LiteralExpression(value, type, b, maxLength, scale, sortOrder, determinism);
-    }
 
-    public LiteralExpression() {
-    }
-    
-    public LiteralExpression(byte[] byteValue) {
-        this.byteValue = byteValue!=null ? byteValue : ByteUtil.EMPTY_BYTE_ARRAY;
-        this.determinism = Determinism.ALWAYS;
-    }
+        /**
+         * Build method for Literal Expressions that have only fields value and determinism (can be null)
+         * or only field byteValue
+         * @param byteValueOnly true if LiteralExpression's only field is byteValue; false otherwise
+         * @return LiteralExpression object
+         */
+        public LiteralExpression buildValueAndDeterminism(boolean byteValueOnly) {
+            if (byteValueOnly) {
+                return new LiteralExpression(null, null, this.byteValue = this.byteValue != null
```

Comment:
I see in PTableImpl, there's a public build method that calls the private initDerivedAttributes method based on whether a particular field is null.
I could change my code to do a similar thing, but in my case the public build method would have no function other than calling the method that initializes derived attributes, since it needs to be called in all cases, if that makes sense.
Either way, the actual LiteralExpression constructor I included is private, and the two build methods (build() and buildValueAndDeterminism(), which call the constructor) are public.
Let me know if you feel I should change it so there's a public build method calling the private initDerivedAttributes method.